File: src/ai_multi_reference_timekeeping/characterization.py
"""Sensor signal characterization helpers."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class RunningStats:
    """Online mean/variance estimator."""

    mean: float = 0.0
    m2: float = 0.0
    count: int = 0

    def update(self, value: float) -> None:
        self.count += 1
        delta = value - self.mean
        self.mean += delta / self.count
        delta2 = value - self.mean
        self.m2 += delta * delta2

    @property
    def variance(self) -> float:
        if self.count < 2:
            return 0.0
        return self.m2 / (self.count - 1)

    @property
    def stddev(self) -> float:
        return self.variance**0.5


class SensorCharacterization:
    """Track reference residuals to characterize sensor noise."""

    def __init__(self) -> None:
        self._stats: dict[str, RunningStats] = {}

    def update(self, name: str, residual: float) -> None:
        stats = self._stats.setdefault(name, RunningStats())
        stats.update(residual)

    def z_score(self, name: str, residual: float) -> float:
        stats = self._stats.get(name)
        if stats is None or stats.count < 2 or stats.stddev == 0:
            return 0.0
        return (residual - stats.mean) / stats.stddev

File: cpp/src/characterization.cpp
#include "aimrt/characterization.hpp"

#include <cmath>

namespace aimrt {

void RunningStats::update(double value) {
    count += 1;
    const double delta = value - mean;
    mean += delta / count;
    const double delta2 = value - mean;
    m2 += delta * delta2;
}

double RunningStats::variance() const {
    if (count < 2) {
        return 0.0;
    }
    return m2 / (count - 1);
}

double RunningStats::stddev() const {
    return std::sqrt(variance());
}

void SensorCharacterization::update(const std::string& name, double residual) {
    stats_[name].update(residual);
}

double SensorCharacterization::z_score(const std::string& name, double residual) const {
    auto it = stats_.find(name);
    if (it == stats_.end()) {
        return 0.0;
    }
    const auto& stats = it->second;
    if (stats.count < 2 || stats.stddev() == 0.0) {
        return 0.0;
    }
    return (residual - stats.mean) / stats.stddev();
}

}  // namespace aimrt
